grabbit/core.py

```diff
@@ -1,26 +1,164 @@
 import json
+import logging
 import os
 import re
+import shutil
+import sys
 from collections import defaultdict, OrderedDict, namedtuple
 from grabbit.external import six, inflect
-from grabbit.utils import natural_sort
-from os.path import join, basename, dirname, abspath, split
+from grabbit.utils import natural_sort, splitext
+from os.path import (join, basename, dirname, abspath, split, exists, islink,
+                     isabs, isdir)
 from functools import partial
+from six import string_types
 
 
 __all__ = ['File', 'Entity', 'Layout']
 
 
 class File(object):
 
-    def __init__(self, filename):
+    def __init__(self, filename, path_patterns=None):
         """
         Represents a single file.
         """
         self.path = filename
         self.filename = basename(self.path)
         self.dirname = dirname(self.path)
         self.entities = {}
+        self.path_patterns = path_patterns
+
+    def replace_entities(self, pattern):
+        new_path = pattern
+        ents = re.findall('\{(.*?)\}', pattern)
+        ents_matched = True
+        for ent in ents:
+            if ent in self.entities:
+                new_path = new_path.replace('{%s}' % ent, self.entities[ent])
+            else:
+                # An entity in the pattern is not an entity for this file
+                ents_matched = False
+
+        if ents_matched:
+            return new_path
+        else:
+            return None
+
+    def get_path(self, path_patterns=None):
```

Comment:
I'm thinking maybe we should call this `build_path` or `construct_path`, to make clear that it potentially returns something different from the existing path. I worry about the situation where someone thinks they're just asking for the existing path (i.e., `File.path`), but because `self.path_patterns` exists, they're going to get something different. A `build_path` method would be unambiguous in that sense (i.e., it would never return the existing `.path`, and would require `path_patterns` to be set somewhere). If we do that, we can either add a `get_path` that's just a getter for `.path`, or we could just leave it as-is and users can call `.path` to get the existing path.